otherlibs/runtime_events/runtime_events.ml

```diff
@@ -237,11 +239,43 @@ module User = struct
 
   external user_register : string -> tag option -> 'a Type.t -> 'a t
                          = "caml_runtime_events_user_register"
-  external user_write : 'a t -> 'a -> unit = "caml_runtime_events_user_write"
+  external user_write : bytes -> 'a t -> 'a -> unit = "caml_runtime_events_user_write"
 
   let register name tag typ = user_register name (Some tag) typ
 
-  let write event value = user_write event value
+  let with_write_buffer =
+    (* [caml_runtime_events_user_write] needs a write buffer in which
+       the user-provided serializer will write its data. The user may
+       want to write a lot of custom events fast, so we want to cache
+       the write buffer across calls.
+
+       To be safe for multi-domain programs, we use domain-local
+       storage for the write buffer. To accomodate for multi-threaded
+       programs (without depending on the Thread module), we store
+       a list of caches for each domain. This might leak a bit of
+       memory: the number of buffers for a domain is equal to the
+       maximum number of threads that requested a buffer concurrently,
+       and we never free those buffers. *)
+    let create_buffer () = Bytes.create 1024 in
+    let write_buffer_cache = Domain.DLS.new_key (fun () -> ref []) in
+    fun consumer ->
+      let buffers = Domain.DLS.get write_buffer_cache in
+      let buf =
+        match !buffers with
+        | [] -> create_buffer ()
+        | b::bs -> buffers := bs; b
+      in
+      Fun.protect ~finally:(fun () -> buffers := buf :: !buffers)
+        (fun () -> consumer buf)
```

Comment:
So if I understand correctly your intent:

- This implementation is not vulnerable to #12677 because the worst that can happen is that, if two threads on the domain run `Domain.DLS.get` concurrently, _and_ this is the first time `Domain.DLS.get` is being run on that domain, then they will get different mutable cells; but this will only cause the drop of an allocated cache, which merely causes unnecessary allocation next time, but no functional bug;
- Thread switches may happen on safe points such as the allocations in `create_buffer ()` and `buf :: !buffers`, and as a result elements of the list can be dropped. But likewise, this causes at worst an additional buffer allocation (and we expect it to be rare).

Is it an accurate description of why you think the code is thread-safe?